### PR TITLE
RFC: Use invokelatest in jl_Function_call

### LINKED ISSUE #343: failure on 0.6 in anonymous callback
In Julia 0.6 master, I can successfully convert an ordinary function like `sum` and call it from Python: `PyObject(sum)(3)` works (and returns 3).

However, the same code no longer works for anonymous functions: `PyObject(x -> x+1)(3)` fails.  I inserted some `println` statements into the [callback hook](https://github.com/JuliaPy/PyCall.jl/blob/16a1bc1adfbe3664819c8c27ade8b3280c106d3e/src/callback.jl#L30-L48), and I verified that `f = unsafe_pyjlwrap_to_objref(self_)::Function` is successfully getting a reference to `f = x -> x+1` (`display(f)` gives the expected output).   However, for some reasons `f(3)` is then throwing a `MethodError`.

@vtjnash, @yuyichao, has something changed at a low-level with `Function` objects, e.g. the new dependency-tracking stuff, that would break this?

(I'm stashing a `jl_value_t*` pointer in the Python object and pulling it out in the callback to call; a reference to the function object is saved in a global dictionary to prevent it from being garbage-collected.)

### PR DESCRIPTION
**Pull request by ihnorton**
Fixes https://github.com/JuliaPy/pyjulia/issues/105

I thought it would also fix #343, but the example in that issue description seems to work fine for me on 0.6 release without this PR (but the issue is still open).

Directly calls `Core._apply_latest` with a closure over kw args as in https://github.com/JuliaLang/julia/pull/22646 for 0.6.

cc @TsurHerman

### CODE REVIEW

**Comment by Keno** [comment]
So @vtjnash should correct me here, but I think this is the wrong place to do it. Instead, pyjulia should update the worldcounter in the ptls whenever it has imported some code. I'm just not sure if there's currently a C Api for that.

File: src/callback.jl [hunk 1]
```diff
@@ -8,7 +8,7 @@
 ################################################################
 
 # Define a Python method/function object from f(PyPtr,PyPtr)::PyPtr.
-# Requires f to be a top-level function.
```
Comment (Keno) [inline src/callback.jl]: typo

**Comment by Keno** [comment]
On the other hand, I also don't really see a downside to doing this here.

File: src/callback.jl [hunk 2]
```diff
@@ -32,11 +32,16 @@ function jl_Function_call(self_::PyPtr, args_::PyPtr, kw_::PyPtr)
     try
         f = unsafe_pyjlwrap_to_objref(self_)::Function
         if kw_ == C_NULL
-            ret = PyObject(f(convert(PyAny, args)...))
```
Comment (Keno) [inline src/callback.jl]: This package still works on 0.5, so you need a check here e.g. `isdefined(Base, :invokelatest)`.

**Comment by vtjnash** [comment]
`jl_Function_call` is already wrapped in a trampoline (via cfunction) which does the equivalent of `invokelatest`, so this shouldn't be necessarily (unless https://github.com/JuliaLang/julia/pull/20167 isn't doing it's job correctly)

**Comment by ihnorton** [comment]
Updated.

@vtjnash I don't know whether that PR is doing its job, but there's a simple repro here: https://github.com/JuliaPy/pyjulia/issues/105#issuecomment-316405459

**Comment by Keno** [comment]
Ok, so the underlying problem here will be addressed by https://github.com/JuliaLang/julia/pull/22987. However, we should probably still merge this here to fix this on current versions of 0.5. Once base is merged, we can put in an appropriate version check.

**Comment by stevengj** [comment]
Thanks!

**Pull request merged**